worker/bundle_service_client.py

```diff
@@ -0,0 +1,247 @@
+            'POST', self._worker_url_prefix(worker_id) + '/reply/' + str(socket_id),
+            data=message)
+
+    @authorized
+    @wrap_exception('Unable to reply to message from bundle service')
+    def reply_data(self, worker_id, socket_id, header_message, fileobj_or_string):
+        method = 'POST'
+        url = self._worker_url_prefix(worker_id) + '/reply_data/' + str(socket_id)
+        query_params = {
+            'header_message': json.dumps(header_message),
+        }
+        if isinstance(fileobj_or_string, basestring):
+            self._make_request(method, url, query_params,
+                               headers={}, data=fileobj_or_string)
+        else:
+            self._upload_with_chunked_encoding(
+                method, url, query_params, fileobj_or_string)
+
+    @authorized
+    @wrap_exception('Unable to start bundle in bundle service')
+    def start_bundle(self, worker_id, uuid, request_data):
+        return self._make_request(
+            'POST', self._worker_url_prefix(worker_id) + '/start_bundle/' + uuid,
+            data=request_data)
+
+    @authorized
+    @wrap_exception('Unable to update bundle metadata in bundle service')
+    def update_bundle_metadata(self, worker_id, uuid, new_metadata):
+        self._make_request(
+            'PUT', self._worker_url_prefix(worker_id) + '/update_bundle_metadata/' + uuid,
+            data=new_metadata)
+
+    @authorized
+    @wrap_exception('Unable to update bundle contents in bundle service')
+    def update_bundle_contents(self, worker_id, uuid, path):
+        with closing(tar_gzip_directory(path)) as fileobj:
+            self._upload_with_chunked_encoding(
+                'PUT', self._worker_url_prefix(worker_id) + '/update_bundle_contents/' + uuid,
+                query_params={'filename': 'bundle.tar.gz'}, fileobj=fileobj)
+
+    @authorized
+    @wrap_exception('Unable to finalize bundle in bundle service')
+    def finalize_bundle(self, worker_id, uuid, request_data):
+        self._make_request(
+            'POST', self._worker_url_prefix(worker_id) + '/finalize_bundle/' + uuid,
+            data=request_data)
+
+    @wrap_exception('Unable to get worker code')
+    def get_code(self):
+        return self._make_request(
+            'GET', '/worker/code.tar.gz', return_response=True, authorized=False)
+
+    @authorized
+    @wrap_exception('Unable to get bundle contents from bundle service')
+    def get_bundle_contents(self, uuid, path):
+        """
+        Returns a file-like object and a file name.
+        """
+        response = self._make_request(
+            'GET', '/bundle/' + uuid + '/contents/blob/' + path,
+            headers={'Accept-Encoding': 'gzip'}, return_response=True)
+        match = re.match('filename="(.*)"',
+                         response.headers['Content-Disposition'])
+        return (response, match.group(1))
+
+    def _make_request(self, method, url, query_params=None, headers={}, data=None,
```

Comment:
`headers={}`
Default arguments are only evaluated once when the function is defined so changes to this `{}` will accidentally persist for future calls.
http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments